modules/firstPartyData/index.js

```diff
@@ -0,0 +1,327 @@
+
+    if (typeBool || !mapping) return key;
+
+    utils.logWarn(`Filtered ${parent}${key} property in ortb2 data: expected type ${(mapping.isArray) ? 'array' : mapping.type}`);
+  }).reduce((result, key) => {
+    let mapping = utils.deepAccess(ORTB_MAP, path + key);
+    let modified = {};
+
+    if (mapping) {
+      if (mapping.optoutApplies && optout) {
+        utils.logWarn(`Filtered ${parent}${key} data: pubcid optout found`);
+        return result;
+      }
+
+      modified = (mapping.type === 'object' && !mapping.isArray)
+        ? validateFpd(fpd[key], path + key + '.children.', parent + key + '.')
+        : (mapping.isArray && mapping.childType)
+          ? filterArrayData(fpd[key], { type: mapping.childType, isArray: mapping.childisArray }, path + key, parent + key) : fpd[key];
+
+      // Check if modified data has data and return
+      (!isEmptyData(modified)) ? result[key] = modified
+        : utils.logWarn(`Filtered ${parent}${key} property in ortb2 data: empty data found`);
+    } else {
+      result[key] = fpd[key];
+    }
+
+    return result;
+  }, {}));
+
+  // Return validated data
+  return validObject;
+}
+
+/**
+ * Resets global ortb2 data
+ */
+export const resetOrtb2 = () => { ortb2 = {} };
+
+function runEnrichments(shouldSkipValidate) {
+  setReferer();
+  setPage();
+  setDomain();
+  setDimensions();
+  setKeywords();
+
+  if (shouldSkipValidate) config.setConfig({ ortb2: utils.mergeDeep({}, ortb2, config.getConfig('ortb2') || {}) });
+}
+
+function runValidations() {
+  let conf = utils.mergeDeep({}, ortb2, validateFpd(config.getConfig('ortb2')));
+
+  config.setConfig({ ortb2: conf });
+
+  let bidderDuplicate = { ...config.getBidderConfig() };
+
+  Object.keys(bidderDuplicate).forEach(bidder => {
+    let modConf = Object.keys(bidderDuplicate[bidder]).reduce((res, key) => {
+      let valid = (key !== 'ortb2') ? bidderDuplicate[bidder][key] : validateFpd(bidderDuplicate[bidder][key]);
+
+      if (valid) res[key] = valid;
+
+      return res;
+    }, {});
+
+    if (Object.keys(modConf).length) config.setBidderConfig({ bidders: [bidder], config: modConf });
+  });
+}
+
+/**
+ * Sets default values to ortb2 if exists and adds currency and ortb2 setConfig callbacks on init
+ */
+export function init() {
+  // Checks for existsnece of pubcid optout cookie/storage
+  // if exists, filters user data out
+  optout = (STORAGE.cookiesAreEnabled() && STORAGE.getCookie('_pubcid_optout')) ||
+    (STORAGE.hasLocalStorage() && STORAGE.getDataFromLocalStorage('_pubcid_optout'));
+  let conf = config.getConfig('firstPartyData');
+  let skipValidations = (conf && conf.skipValidations) || false;
+  let skipEnrichments = (conf && conf.skipEnrichments) || false;
+
+  if (!skipEnrichments) runEnrichments(skipValidations);
+  if (!skipValidations) runValidations();
+}
+
+function addBidderRequestHook(fn, bidderRequests) {
+  init();
+  resetOrtb2();
```

Comment:
Not sure this is right -- what if the publisher adds their own ortb2 data from the page? This module needs to preserve it and merge in the enrichments.

Reply:
This reset is for the ortb2 object within the module itself. Not the global one used in setConfig(...). This is to allow enrichments to be properly reevaluated on a subsequent auction if refreshFPD is triggered prior